lib/core/util.js

```diff
@@ -26,6 +26,57 @@ function isBlobLike (object) {
   )
 }
 
+function encode (val) {
+  return encodeURIComponent(val).replace(/%3A/gi, ':').replace(/%24/g, '$').replace(/%2C/gi, ',').replace(/%20/g, '+').replace(/%5B/gi, '[').replace(/%5D/gi, ']')
+}
+
+function isDate (val) {
+  return toString.call(val) === '[object Date]'
+}
+
+function isObject (val) {
+  return val !== null && typeof val === 'object'
+}
+
+function buildURL (url, params) {
+  if (!params) {
+    return url
+  }
+
+  const parts = []
+  for (let [key, val] of Object.entries(params)) {
+    if (val === null || typeof val === 'undefined') {
+      return
+    }
+
+    if (!Array.isArray(val)) {
+      val = [val]
+    }
+
+    for (let v of val) {
+      if (isDate(v)) {
+        v = v.toISOString()
+      } else if (isObject(v)) {
+        v = JSON.stringify(v)
+      }
+      parts.push(encode(key) + '=' + encode(v))
+    }
+  }
+
+  const serializedParams = parts.join('&')
+
+  if (serializedParams) {
+    const hashmarkIndex = url.indexOf('#')
+    if (hashmarkIndex !== -1) {
+      url = url.slice(0, hashmarkIndex)
+    }
+
+    url += (url.indexOf('?') === -1 ? '?' : '&') + serializedParams
+  }
+
+  return url
+}
```

Comment:
This is getting too complicated. What was wrong with the previous version with URLSearchParams?

Reply:
It did not handle all of the relevant cases: e. g. arrays, dates or objects: implicit toString() is not a particularly good solution.

I'm happy to drop any parts of this logic that you believe to be complicated; I've already replaced the most opinionated part (object serialization) and happy to remove anything else that you believe to be excessive (e. g. hashmark handling)

Reply:
I'd prefer to just do the exact same thing that `URLSearchParams` does.

Reply:
@ronag It is significantly less useful than the current implementation, especially with regards to the handling of arrays, and is going to be a better option in no cases: no framework is capable of automatically parsing `[1,2]` as an array, while current implementation will be automatically decoded by pretty much anything with the support of arrays.

@targos @mcollina Could you please provide a second/third opinion?